caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -316,71 +316,113 @@ public boolean isFile( ConcreteResource resource )
     public InputStream openInputStream( final ConcreteResource resource )
             throws IOException
     {
-        //use "this" as lock is heavy, should think about use the transfer for the resource as the lock for each thread
+        final String pathKey = getKeyForResource( resource );
+
+        // This lock is used to control the the local resource can be opened successfully finally when local resource missing
+        // but NFS not, which means will do a NFS->local copy.
+        final Object copyLock = new Object();
+        // A flag to mark if the local resource can be open now or need to wait for the copy thread completes its work
+        final AtomicBoolean canStreamOpen = new AtomicBoolean( false );
```

Comment:
Dumb idea maybe, but could you use this as the lock object instead of creating a separate Object? Then, I guess you'd just notifyAll() / wait() for this instead